### PR TITLE
[FEATURE] [Driver-3] Simplified load-balancing policy, improved logging, better test coverage, and SSR-informed retry policy

### PR DESCRIPTION
**Pull request by David-Noble-at-work**
This change:

- Updates the package and module names for consistency with Azure SDK conventions. This is release 1.0.0. Now is the time to make this change. The new names are:

  - package: `azure-cosmos-cassandra-driver-3-extensions`
  - module: `com.azure.cosmos.cassandra`

- Simplifies `CosmosLoadBalancingPolicy`. It is now based on the specification of just two settings that make clear how regional failover works:
  ```java
  CosmosLoadBalancingPolicy.builder()
      .withMultiRegionWrites(enabled /* default:false */)
      .withPreferredRegions(Arrays.asList("region-1", "region-2", "etc.")
      .build();
  ```
  We assume that the global endpoint is the one-and-only one contact point. We failover from the first to the last region in the list of preferred regions in the order that they're listed. If all those regions are down, we failover to the region hosting the global endpoint. If that region is down, we failover to any remaining regions that are available in lexical order.

  If multi-region writes are disabled, we only ever write to the global endpoint. We rely on server-side failover rules in this case. 

  We also attempt to handle Apache Cassandra usage scenarios with some grace. We assume that all contact points are writable. You may write to any of the regions specified in that list. The behavior is otherwise the same.

- Updates `CosmosRetryPolicy` with behavior that's informed by the forthcoming Server-Side Retry feature.
   Most notably you can now enable or disable read/write retries.
  ```java
  CosmosRetryPolicy.builder()
      .withReadTimeoutRetries(enabled /* default: true */)
      .withWriteTimeoutRetries(enabled /* default: false */)
      .build();
   ```

 This spreadsheet documents the new retry policy, comparing it to the `DefaultRetryPolicy`: [CosmosRetryPolicy.pdf](https://github.com/Azure/azure-cosmos-cassandra-extensions/files/6695995/CosmosRetryPolicy.pdf)

- Significantly Improves logging.
 
  I moved, extended, and simplified the Json serializers that were defined in `CosmosLoadBalancingPolicy`. All Json serializers now live  in `com.azure.cosmos.cassandra.implementation.serializer`. I use them extensively in debug/trace log messages to gain deeper insights into the state of the driver and extensions at run time.

  There are about a dozen files. The  implementations are trivial. Most are metadata extensions that simply order, ignore, or include properties. Adding and removing serializers is trivial. See `JsonRegistrar`. The code Is now well tested. See `JsonTest`.

- Significantly improves test coverage. See the test code. We now interrogate the heck out of `CosmosLoadBalancingPolicy` and make a more conscientious effort to test OverloadedException retries. See `CosmosRetryPolicy.canRetryAsExpectedWhenThrottled`. 

  There's a bit of work to do on that method. I'll leave that for a follow-on chore. It's currently excluded from CI.

### CODE REVIEW

**Review by TheovanKraay — APPROVED** [review]

File: azure-pipelines.yml [hunk 1]
```diff
@@ -12,44 +12,51 @@ pr:
...
...MDWAezSmHXdX17YULePcrWNSkcULzmXw5xRWbLPd7M7IHLfKYbuCB06PZ3rmLyM13LbQrzsMEeNDuaTu4aGSP95CJiIiAVt9jZ6ZSWw3FHm8mTvs0WE2JXRC3SRzLuk5vSAtMz2iKQ4EscejDGcyyKFsMkpmBdIsT53UEP9CCfqywht5G7v6AgHrWPeZIYqiZc2+j4JhfjkcqxdreEXaL6lIl1fvZq1dq+nLWNgVrIh/BsWxjhkIS2zuoqx8PQra2rXsZX2iDQj6hQ/1aOfuUqvXHabfST8+xgdqP0U8sP29Z4mlXo+xhSN9yIb+r9bp2QMmGbgzxpU86SnSMiJwoGCp/r9U10jLVxIjtL1NqlpL3ouCFYC6FXBWi8MW2ZDEYJ3gB5WlSVHXBX5+o5Wu4LE947YVwRWDS2sdt26tUYw2GHiO0SlUYvc3bSCtK9poDuFY0f7ufsaz8QyXDooADxIM40e3PaCSg/YvItBMaOFaVHHzaGNEt+pmrcohvJVIgXF7B4SELcCnC2Moye8yGpGy+ODnocFcuGK6rkjudElIFIfbI/9NKCtzcen41b+69D/CpxX3yaEZPw1Tb6+80X57Zozy0WIGfv6mTADD7GS3dG5pcW59y5cG/aVZcTBlWyB2TECyBf5Bm8iHyeSMdUkKjHSisoRdP0OWXXQI9Xxvy40QkmmcagN5fEcwVnZjsRcMDB2p9FqiX6aGOXdEeIvIgNLHoDjw1HCMqVO4UAuZ74dyLmHAn+Jc5393DWpH5sSQbcMuZ/3O9FqCRt94qG1Wbnp/DUTsy71OiH11M9K9HSGwkVbwiz1oehMt42/IBhJ/2WrJ6lldflD7o+3RtroDcR6cb8oBYqSjoMJOUdxzzLzwqBrQynLKAIvRFEE4CA9A7auRzy9nu+EyVdc2E+OwIB8ymVa16K/OjRK+hIClF80t5O41iUz6019hyLb39Bk8T0oEPv8naNnUhb2gqPBcHWastSXnpdLA4jeUwbp60X0jgw9k+aQuYJHeX8Tix0EMIp0wcBDC7Sp7CEmd2Rvpovtx1w51yadDZ1IsWgxAaCByGz6KibwwOpxPHAfPo0P9BBtdBqPTYAIc6t280CLMD4wITAJBgUrDgMCGgUABBQTVTI0xnCLmO1UYS1iD9ZPVhaeuQQU0b9I3cUSd9/VfpENPWDQ4kn8FosCAwGGoA=='
```
Comment (vivekr20) [inline azure-pipelines.yml]: Why is this needed?

File: driver-3/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ## Release History
 
-## 0.14.0
+## 1.0.0-SNAPSHOT
```
Comment (vivekr20) [inline driver-3/CHANGELOG.md:3]: does it still need the -SNAPSHOT suffix?

File: driver-3/src/main/java/com/azure/cosmos/cassandra/CosmosLoadBalancingPolicy.java [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+import java.util.stream.Collectors;
+
+import static com.azure.cosmos.cassandra.implementation.Json.toJson;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Implements a {@link LoadBalancingPolicy} with an option to specify read datacenter and write datacenter to route read
+ * and write requests to their corresponding data centers.
+ * <p>
+ * If a {@code readDC} is specified, we prioritize nodes in the read datacenter for read requests. Either one of {@code
```
Comment (vivekr20) [inline driver-3/src/main/java/com/azure/cosmos/cassandra/CosmosLoadBalancingPolicy.java:49]: These comments need to be updated to reflect using PreferredRegions instead?

File: driver-3/src/main/java/com/azure/cosmos/cassandra/CosmosLoadBalancingPolicy.java [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+         * These regions will appear last in the list of preferred regions following those specified using
+         * {@link CosmosLoadBalancingPolicy#builder()}.
+         * <p>
+         * This method is not thread safe and can only be called once. It should only be called by {@link #init}.
+         *
+         * @param contactPoints A set of contact points.
+         *
+         * @throws IllegalStateException if this method is called more than once.
+         */
+        void addPreferredRegions(final List<Host> contactPoints) {
```
Comment (vivekr20) [inline driver-3/src/main/java/com/azure/cosmos/cassandra/CosmosLoadBalancingPolicy.java]: nit, should this be marked private? (I'm guessing that's the default?)

**Review by vivekr20 — APPROVED** [review]

File: driver-3/CHANGELOG.md (hunk 2 above)
Comment (David-Noble-at-work) [reply driver-3/CHANGELOG.md:3]: My process is to bump the version number from the develop branch.

File: azure-pipelines.yml (hunk 1 above)
Comment (David-Noble-at-work) [reply azure-pipelines.yml]: It's good practice to rely on a trust store when hostname validation is turned off as it must be to access multi-region Cassandra API instances. Driver-3 turns hostname validation off by default. Driver-4 turns hostname validation on by default. We choose to use turn hostname validation off and use a trust store.
